### PR TITLE
feat: add optional Prof rpc

### PR DESCRIPTION
**Pull request by RetricSu**
we have new relic to watch for perf on web3. 

This pr adds Prof RPC module to enable quick debugging on the local environment. Only with `ENABLE_PROF_RPC=true` will allow such a module. like [ckb](https://github.com/nervosnetwork/ckb/blob/develop/rpc/src/service_builder.rs#L190-L196)

- `prof_cpu` to debug CPU profiling
- `prof_heap` to debug memory usage/ memory leak

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,6 +37,7 @@ MAX_QUERY_NUMBER=<optional, integer number, maximum number of records to be retu
 MAX_QUERY_ROUNDS=<optional, integer number, maximum number of round trips to database done by a single query for records>
 MAX_QUERY_TIME_MILSECS=<optional, integer number, maximum number of time for database query>
 FEE_RATE=<optional, decimal number, use to calculate minimal l2tx fee required, default to 0>
+ENABLE_DEBUG_RPC=<optional, boolean, default to false>
```
Comment (keroro520) [inline README.md:40]: It mainly focuses on profiling. What about renaming all "debug" to "profile".

File: packages/api-server/src/decorator.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { asyncSleep } from "./util";
+import path from "path";
+import fs from "fs";
+
+const Profiler = require("v8-profiler-node8");
```
Comment (classicalliu) [inline packages/api-server/src/decorator.ts:5]: Any problem with https://github.com/hyj1991/v8-profiler-next ?

Seems https://github.com/hyj1991/v8-profiler-node8/blob/master/v8-profiler.js#L1 only require `v8-profiler-next` and do nothing more ?

And `v8-profiler-next` has typescript support

File: packages/api-server/src/methods/modules/index.ts [hunk 3]
```diff
@@ -1,7 +1,15 @@
-export const list = ["Eth", "Web3", "Net", "Gw", "Poly"];
+import { envConfig } from "../../base/env-config";
+
+let enableList = ["Eth", "Web3", "Net", "Gw", "Poly"];
```
Comment (classicalliu) [inline packages/api-server/src/methods/modules/index.ts:3]: `const` also works

**Review by classicalliu — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (RetricSu) [reply README.md:40]: will update rpc method to `prof_cpu` and `prof_memory`, env var to `ENABLE_PROF_RPC `

File: packages/api-server/src/decorator.ts (hunk 2 above)
Comment (RetricSu) [reply packages/api-server/src/decorator.ts:5]: seems the author rewrite `v8-profiler-node8` with `v8-profiler-next`, wll replace with `v8-profiler-next`

File: README.md (hunk 1 above)
Comment (RetricSu) [reply README.md]: updated

File: packages/api-server/package.json [hunk 4]
```diff
@@ -61,15 +61,18 @@
...
     "@types/express-ws": "^3.0.1",
+    "@types/heapdump": "^0.3.1",
     "@types/http-errors": "^1.8.2",
     "@types/leveldown": "^4.0.3",
     "@types/levelup": "^4.3.3",
     "@types/newrelic": "^7.0.2",
     "@types/secp256k1": "^4.0.2",
     "ava": "^4.1.0",
     "concurrently": "^6.0.0",
+    "heapdump": "^0.3.15",
```
[the remark this answers is not in the document]
Comment (RetricSu) [reply packages/api-server/package.json]: updated

File: packages/api-server/package.json [hunk 5]
```diff
@@ -61,15 +61,18 @@
...
     "nodemon": "^2.0.7",
     "ts-node": "^9.1.1",
-    "typescript": "^4.2.2"
+    "typescript": "^4.2.2",
+    "v8-profiler-node8": "^6.4.0"
```
[the remark this answers is not in the document]
Comment (RetricSu) [reply packages/api-server/package.json]: updated

File: packages/api-server/src/decorator.ts (hunk 2 above)
Comment (RetricSu) [reply packages/api-server/src/decorator.ts]: updated

File: packages/api-server/src/methods/modules/index.ts (hunk 3 above)
Comment (RetricSu) [reply packages/api-server/src/methods/modules/index.ts]: updated

**Review by classicalliu — APPROVED** [review]

File: packages/api-server/src/methods/modules/prof.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        transform.on("finish", () => {
+          snapshot.delete.bind(snapshot);
+          resolve(fileName);
+        });
+        transform.on("error", reject);
+      });
+    };
+    const name = path.join(`${Date.now()}.heapsnapshot`);
+    await createHeadDumpFile(name);
+    return "ok";
```
Comment (keroro520) [inline packages/api-server/src/methods/modules/prof.ts:32]: What about returning the file path of the heap snapshot?

File: packages/api-server/src/decorator.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        await asyncSleep(timeMs);
+      }
+      const profile = v8Profiler.stopProfiling();
+      const cpuprofile = path.join(`${Date.now()}.cpuprofile`);
+      profile
+        .export()
+        .pipe(fs.createWriteStream(cpuprofile))
+        .on("finish", () => profile.delete());
+
+      return result;
```
Comment (keroro520) [inline packages/api-server/src/decorator.ts:30]: What about returning the file path of the CPU profile?

And I am curious about what the `result` variable looks like?

**Review by keroro520 — CHANGES_REQUESTED** [review]

File: packages/api-server/src/decorator.ts (hunk 7 above)
Comment (RetricSu) [reply packages/api-server/src/decorator.ts:30]: result is the function execute result which decoder wrapper at. nothing to do with CPU profile here

File: packages/api-server/src/methods/modules/prof.ts (hunk 6 above)
Comment (RetricSu) [reply packages/api-server/src/methods/modules/prof.ts:32]: I am not so sure. assume the RPC module is accidentally open in production, this file path might expose some info ? the file will be located in the root project by the way.

Comment (keroro520) [reply packages/api-server/src/methods/modules/prof.ts:32]: I don't think the file path would expose any secret info.
Or returns the filename only?
Explicit file names are helpful when debugging.

Comment (RetricSu) [reply packages/api-server/src/methods/modules/prof.ts]: return filename will be great, updated

**Review by keroro520 — APPROVED** [review]

**Pull request merged**